Replace `Handle::Weak` with `Handle::Uuid`. (#19896)
bevyengine/bevy: release-content/migration-guides/handle_weak_replaced_with_handle_uuid.md

---
title: `Handle::Weak` has been replaced by `Handle::Uuid`.
pull_requests: [19896]
---

> `Handle::Weak` has some weird behavior. It allows for a sprite to be given a handle that is dropped
> **while the sprite is still using it**. This also results in more complexity in the asset system.
> The primary remaining use for `Handle::Weak` is to store asset UUIDs initialized through the
> `weak_handle!` macro. To address this, `Handle::Weak` has been replaced by `Handle::Uuid`!

Reviewer: ```suggestion
`Handle::Weak` had some weird behavior. It allowed for a sprite to be given a handle that was dropped
**while the sprite is still using it**. This also resulted in more complexity in the asset system.
The primary remaining use for `Handle::Weak` was to store asset UUIDs initialized through the
`weak_handle!` macro. To address this, `Handle::Weak` has been replaced by `Handle::Uuid`!
```

Author: Agreed, save for the last "was", since our primary remaining use for `Handle::Weak` is still present, so we have to consider it. I'm keeping that one present tense.

Users using the `weak_handle!` macro should switch to the `uuid_handle!` macro.

```rust

> # Before

Author: No wonder this wasn't highlighting correctly LMAO. Done!

> const IMAGE: Handle<Image> = weak_handle!("12345678-9abc-def0-1234-56789abcdef0");

Reviewer: ```suggestion
const IMAGE: Handle<Image> = weak_handle!("b20988e9-b1b9-4176-b5f3-a6fa73aa617f");
```
Let's use a real UUID instead of a synthetic one here, otherwise a reader might get the impression they are meant to smash their keyboard manually ;)

Author: Oh you're supposed to generate them? I would be surprised if users actually did that hahaha.

I don't care enough, so done.

# After

> const IMAGE: Handle<Image> = uuid_handle!("12345678-9abc-def0-1234-56789abcdef0");

Reviewer: ```suggestion
const IMAGE: Handle<Image> = uuid_handle!("b20988e9-b1b9-4176-b5f3-a6fa73aa617f");
```

```

Users using the `Handle::Weak` variant directly should consider replacing it with `AssetId` instead,

> accessible through `Handle::id`. These situations are very case-by-case migrations.

Author: I don't think one really exists? Like the assumption is that users don't need this. I can't think of an actual situation where you'd be using this. That's why I added the "case-by-case" line - there isn't a generally pattern here that we know users are using.

> accessible through `Handle::id`. These situations are very case-by-case migrations.

Author: Very good point, I totally forgot to include this in the migration guide! Seeing how some of our own code still uses clone_weak by accident, I would be shocked if users weren't also doing this on accident. Calling it out is important here.
